File: frontend/src/app/shared/services/tramite.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import {
  Tramite,
  CreateTramiteRequest,
  AvanzarTramiteRequest,
  FormularioActualResponse,
  TramiteStats,
  Apelacion
} from '../models/tramite.model';

@Injectable({ providedIn: 'root' })
export class TramiteService {
  private readonly base = `${environment.apiUrl}/tramites`;

  constructor(private readonly http: HttpClient) {}

  getAll(
    page = 0,
    size = 10,
    estado?: string
  ): Observable<{ content: Tramite[]; totalElements: number }> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    if (estado && estado.trim() !== '') {
      params = params.set('estado', estado);
    }
    return this.http.get<{ content: Tramite[]; totalElements: number }>(this.base, { params });
  }

  getStats(): Observable<TramiteStats> {
    return this.http.get<TramiteStats>(`${this.base}/stats`);
  }

  getById(id: string): Observable<Tramite> {
    return this.http.get<Tramite>(`${this.base}/${id}`);
  }

  create(req: CreateTramiteRequest): Observable<Tramite> {
    return this.http.post<Tramite>(this.base, req);
  }

  avanzar(id: string, req: AvanzarTramiteRequest): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.base}/${id}/avanzar`, req);
  }

  responder(id: string, observaciones: string): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.base}/${id}/responder`, { observaciones });
  }

  getFormularioActual(id: string): Observable<FormularioActualResponse> {
    return this.http.get<FormularioActualResponse>(`${this.base}/${id}/formulario-actual`);
  }

  tomar(id: string): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.base}/${id}/tomar`, {});
  }

  getTramitesSinAsignar(
    page = 0,
    size = 20
  ): Observable<{ content: Tramite[]; totalElements: number }> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<{ content: Tramite[]; totalElements: number }>(
      `${this.base}/sin-asignar`,
      { params }
    );
  }

  asignarManual(tramiteId: string, funcionarioId: string): Observable<Tramite> {
    return this.http.post<Tramite>(
      `${this.base}/${tramiteId}/asignar-manual`,
      { funcionarioId }
    );
  }

  observar(id: string, motivo: string, documentosIds?: string[]): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.base}/${id}/observar`, { motivo, documentosIds });
  }

  denegar(id: string, motivo: string, documentosIds?: string[]): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.base}/${id}/denegar`, { motivo, documentosIds });
  }

  apelar(id: string, justificacion: string, documentosIds?: string[]): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.base}/${id}/apelar`, { justificacion, documentosIds });
  }

  resolverApelacion(id: string, aprobada: boolean, observaciones?: string): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.base}/${id}/resolver-apelacion`, { aprobada, observaciones });
  }

  getApelacion(id: string): Observable<Apelacion> {
    return this.http.get<Apelacion>(`${this.base}/${id}/apelacion`);
  }
}

File: mobile/lib/data/services/tramite_service.dart
import '../models/campo.dart';
import '../models/tramite.dart';
import 'api_client.dart';

class TramiteService {
  TramiteService({required ApiClient apiClient}) : _api = apiClient;

  final ApiClient _api;

  Future<List<TramiteResponse>> getMisTramites({
    String? estado,
    int page = 0,
  }) async {
    final params = <String, String>{
      'page': '$page',
      'size': '20',
      if (estado != null) 'estado': estado,
    };
    final query = Uri(queryParameters: params).query;
    final path = '/tramites/mis-tramites?$query';
    final resp = await _api.get(path);
    return _pageContent(resp);
  }

  Future<List<TramiteResponse>> getBandeja({
    String? estado,
    int page = 0,
  }) async {
    final params = <String, String>{
      'page': '$page',
      'size': '20',
      if (estado != null) 'estado': estado,
    };
    final query = Uri(queryParameters: params).query;
    final path = '/tramites?$query';
    final resp = await _api.get(path);
    return _pageContent(resp);
  }

  Future<TramiteResponse> getById(String id) async {
    final resp = await _api.get('/tramites/$id');
    return TramiteResponse.fromJson(resp);
  }

  Future<FormularioActualResponse> getFormularioActual(String tramiteId) async {
    final resp = await _api.get('/tramites/$tramiteId/formulario-actual');
    return FormularioActualResponse.fromJson(resp);
  }

  Future<TramiteResponse> crear(String politicaId) async {
    final resp = await _api.post('/tramites', {'politicaId': politicaId});
    return TramiteResponse.fromJson(resp);
  }

  Future<TramiteResponse> avanzar(
    String id, {
    required String accion,
    String? observaciones,
    Map<String, dynamic>? datos,
  }) async {
    final body = <String, dynamic>{
      'accion': accion,
      if (observaciones != null) 'observaciones': observaciones,
      'datos': datos ?? {},
    };
    final resp = await _api.post('/tramites/$id/avanzar', body);
    return TramiteResponse.fromJson(resp);
  }

  Future<TramiteResponse> responder(
    String id, {
    Map<String, dynamic>? datos,
    String? observaciones,
  }) async {
    final body = <String, dynamic>{
      'accion': 'RESPONDER',
      if (observaciones != null) 'observaciones': observaciones,
      'datos': datos ?? {},
    };
    final resp = await _api.post('/tramites/$id/responder', body);
    return TramiteResponse.fromJson(resp);
  }

  Future<TramiteResponse> tomar(String id) async {
    final resp = await _api.post('/tramites/$id/tomar', {});
    return TramiteResponse.fromJson(resp);
  }

  Future<TramiteResponse> observar(
    String id, {
    required String motivo,
  }) async {
    final resp = await _api.post('/tramites/$id/observar', {'motivo': motivo});
    return TramiteResponse.fromJson(resp);
  }

  Future<TramiteResponse> apelar(
    String id, {
    required String justificacion,
  }) async {
    final resp =
        await _api.post('/tramites/$id/apelar', {'justificacion': justificacion});
    return TramiteResponse.fromJson(resp);
  }

  Future<TramiteResponse> denegar(
    String id, {
    required String motivo,
  }) async {
    final resp = await _api.post('/tramites/$id/denegar', {'motivo': motivo});
    return TramiteResponse.fromJson(resp);
  }

  Future<TramiteResponse> resolverApelacion(
    String id, {
    required bool aprobada,
    String? observaciones,
  }) async {
    final body = <String, dynamic>{
      'aprobada': aprobada,
      if (observaciones != null && observaciones.isNotEmpty)
        'observaciones': observaciones,
    };
    final resp = await _api.post('/tramites/$id/resolver-apelacion', body);
    return TramiteResponse.fromJson(resp);
  }

  List<TramiteResponse> _pageContent(Map<String, dynamic> resp) {
    final content = resp['content'] as List<dynamic>? ?? const [];
    return content
        .map((e) => TramiteResponse.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}
